### PR TITLE
Add use of Windows Wide CRT API for env. vars

### LINKED ISSUE #20083: `getEnv` doesn't return UTF-8 on Windows
Hello. :) I'm reporting a regression. Issue is that `getEnv` doesn't return
valid UTF-8 on Windows.

Why do I expect `getEnv` to return UTF-8? Nim expects `getEnv` to return a
UTF-8 string. This is evident in `std/os.findExe`, where it assumes that
`getEnv("PATH")` is UTF-8 encoded, because it calls `fileExists`, which calls
`GetFileAttributesW`, converting portions of PATH from UTF-8 to UTF-16.

Looking at the code, the Windows implementation retrieves the environment
variable from ANSI API. This is `getenv` and `putenv`, which, after
researching, returns strings encoded in the currently or previously used
Windows codepage.

### Example

```nim
import std/envvars, sequtils
proc c_wputenv(envstring: WideCString): cint {.importc: "_wputenv", header: "<stdlib.h>".}
const
  envName = "test13"
  strUtf8 = "\xc3\x86" # "LATIN CAPITAL LETTER AE" in UTF-8
discard c_wputenv(newWideCString(envName & "=" & strUtf8))
echo getEnv(envName).toSeq
```

### Current Output

```
@['\xc6']
```

This is `strUtf8` encoded in my code page (windows-1252).

### Expected Output

```
@['\xc3', '\x86']
```

### Possible Solution

It's not reliable to convert from ANSI encoding because the code page can
change over the course of a process' lifetime.

A solution is to use `_wgetenv` CRT API, which returns the environment variable
in UTF-16 encoding, and Nim can then convert it to UTF-8. I have already
implemented this solution. I will make a PR after submitting this issue.

### Additional Information

My Windows codepage:

```nim
import encodings
echo getCurrentEncoding(true)
````

returns: `windows-1252`

`getEnv` returned UTF-8 on Windows before `6b3c77e` (#18575), so this is a
regression. When bisecting, imported `std/os` instead of `std/envvars`, because
`std/envvars` was added recently.

```
$ nim -v
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2022-07-17
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 0d8bec695606a65c5916d0da7fcb0a976a4e1f7b
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by havardjohn**
Replaces use of CRT API `getenv` and `putenv` with respectively
`_wgetenv` and `_wputenv`. Motivation is to reliably convert environment
variables to UTF-8, and the wide API is best there, because it's
reliably UTF-16.

Changed the hack in `lib/std/private/win_setenv.nim` by switching the
order of the Unicode and MBCS environment update; Unicode first, MBCS
second. Because `_wgetenv`/`_wputenv` is now used, the Unicode
environment will be initialized, so it should always be updated.

Stop updating MBCS environment with the name of `getEnv`. It's not
necessarily true that MBCS encoding and the `string` encoding is the
same. Instead convert UTF-16 to current Windows code page with
`wcstombs`, and use that string to update MBCS.

Fixes regression in `6b3c77e` that caused `std/envvars.getEnv` or
`std/os.getEnv` on Windows to return non-UTF-8 encoded strings.

Add tests that test environment variables with Unicode characters in
their name or value.

Changed `tnewruntime_misc.nim` test to account for the extra allocation
in `getEnv` for Windows.

`osenv.nim` and `envvars.nim` (as well as `tosenv.nim` and
`tenvvars.nim`) are duplicate modules, so changes in this PR are
duplicated as well. I assume these modules will be deduplicated in the
future.

Fixes #20083

### CODE REVIEW

**Comment by havardjohn** [comment]
I'll be gone for 2 weeks, so setting this to draft

**Comment by havardjohn** [comment]
I seem to be running into pitfall after pitfall while trying to get the tests to run. Right now `tests/misc/trunner_special.nim` doesn't run in CI, seemingly due to some SSL error. I cannot reproduce this error locally. It doesn't seem like this PR should affect this test, so my plan is to retry CI in a week or so.

**Comment by ringabout** [comment]
> I seem to be running into pitfall after pitfall while trying to get the tests to run. Right now tests/misc/trunner_special.nim doesn't run in CI, seemingly due to some SSL error. I cannot reproduce this error locally.

fixed by https://github.com/nim-lang/Nim/pull/20181

**Comment by havardjohn** [comment]
Thanks, @ringabout. Undrafting the PR.

File: lib/std/private/win_setenv.nim [hunk 1]
```diff
@@ -57,40 +58,44 @@ else:
...
-      if mbstowcs(buf2, cstring(name), csize_t(requiredSize + 1)) == csize_t(high(uint)):
-        errno = EINVAL
-        return -1
-      var ptrToEnv = cast[WideCString](c_wgetenv(buf2))
-      ptrToEnv[0] = '\0'.Utf16Char
-      ptrToEnv = cast[WideCString](c_wgetenv(buf2))
-      ptrToEnv[1] = '='.Utf16Char
+    if genviron != nil:
+
+      # wcstombs returns (size_t) (-1) if any characters cannot be represented
```
Comment (Varriount) [inline lib/std/private/win_setenv.nim]: I believe this would be more accurate as
```suggestion
      # wcstombs returns `high(csize_t)` if any characters cannot be represented
```

File: lib/std/private/win_setenv.nim [hunk 2]
```diff
@@ -25,27 +25,28 @@ when not defined(windows): discard
...
 
   proc setEnvImpl*(name: string, value: string, overwrite: cint): cint =
     const EINVAL = cint(22)
-    if overwrite == 0 and c_getenv(cstring(name)) != nil: return 0
+    let wideName = name.newWideCString
+    if overwrite == 0 and c_wgetenv(wideName) != nil: return 0
     if value != "":
       let envstring = name & "=" & value
-      let e = c_putenv(cstring(envstring))
+      let e = c_wputenv(envstring.newWideCString)
```
Comment (Varriount) [inline lib/std/private/win_setenv.nim]:
```suggestion
      let e = c_wputenv(newWideCString(envstring))
```

File: lib/std/private/win_setenv.nim [hunk 3]
```diff
@@ -25,27 +25,28 @@ when not defined(windows): discard
...
 
-  proc mbstowcs(wcstr: ptr wchar_t, mbstr: cstring, count: csize_t): csize_t {.importc: "mbstowcs", header: "<stdlib.h>".}
+  proc wcstombs(wcstr: ptr char, mbstr: WideCString, count: csize_t): csize_t {.importc, header: "<stdlib.h>".}
     # xxx cint vs errno_t?
 
   proc setEnvImpl*(name: string, value: string, overwrite: cint): cint =
     const EINVAL = cint(22)
-    if overwrite == 0 and c_getenv(cstring(name)) != nil: return 0
+    let wideName = name.newWideCString
+    if overwrite == 0 and c_wgetenv(wideName) != nil: return 0
```
Comment (Varriount) [inline lib/std/private/win_setenv.nim]:
```suggestion
    let wideName = newWideCString(name)
    if overwrite == 0 and c_wgetenv(wideName) != nil:
      return 0

```

File: lib/std/private/win_setenv.nim [hunk 4]
```diff
@@ -57,40 +58,44 @@ else:
...
+    s[0] = '\0'.Utf16Char
     #[
     This would result in a double null termination, which normally signifies the
     end of the environment variable list, so we stick a completely empty
     environment variable into the list instead.
     ]#
-    s = c_getenv(cstring(name))
-    s[1] = '='
+    s = c_wgetenv(wideName)
+    s[1] = '='.Utf16Char
```
Comment (Varriount) [inline lib/std/private/win_setenv.nim]:
```suggestion
    s[1] = Utf16Char('=')
```

File: lib/std/private/win_setenv.nim [hunk 5]
```diff
@@ -57,40 +58,44 @@ else:
...
     let envstring = name & "=  "
-    if c_putenv(cstring(envstring)) != 0:
+    if c_wputenv(newWideCString(envstring)) != 0:
       errno = EINVAL
       return -1
     # Here lies the documentation we blatently ignore to make this work.
-    var s = c_getenv(cstring(name))
-    s[0] = '\0'
+    var s = c_wgetenv(wideName)
+    s[0] = '\0'.Utf16Char
```
Comment (Varriount) [inline lib/std/private/win_setenv.nim]:
```suggestion
    s[0] = Utf16Char('\0')
```

**Comment by yglukhov** [comment]
There are some new calls to `newWideCString`, which allocates, and then it is not freed, or am I missing something? cc @havardjohn @Varriount

**Comment by Araq** [comment]
The GC cleans up `newWideCString` calls.
